### PR TITLE
linagora/lgs/linshare/products/linshare-mobile-apps#584 LA-584: add share suggestion presentation

### PR DESCRIPTION
**Pull request by manhduc-nguyen**
Screen Shot 2020-12-10 at 18 04 03 Screen Shot 2020-12-11 at 17 26 59

### CODE REVIEW

**Comment by manhduc-nguyen** [comment]
@hoangdat @imGok please review this a bit, thanks

File: lib/presentation/widget/upload_file/upload_file_viewmodel.dart [hunk 1]
```diff
@@ -132,16 +131,50 @@ class UploadFileViewModel extends BaseViewModel {
...
+    if (autoCompleteResult is SimpleAutoCompleteResult) {
+      return autoCompleteResult.identifier;
+    } else if (autoCompleteResult is UserAutoCompleteResult) {
+      return autoCompleteResult.mail.isNotEmpty
+          ? autoCompleteResult.mail
+          : autoCompleteResult.display;
+    } else {
+      return autoCompleteResult.display;
+    }
+  }
```
Comment (hoangdat) [inline lib/presentation/widget/upload_file/upload_file_viewmodel.dart:162]: many place can reuse this method => IMO extensions function is better.

File: lib/presentation/view/avatar/label_avatar_builder.dart [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:linshare_flutter_app/presentation/util/extensions/color_extension.dart';
+
+class LabelAvatarBuilder {
+  final String label;
+  Color _backgroundColor;
+
+  LabelAvatarBuilder(this.label);
```
Comment (imGok) [inline lib/presentation/view/avatar/label_avatar_builder.dart:41]: Maybe you should add `key `aswell to test it later

File: domain/lib/src/usecases/autocomplete/get_autocomplete_sharing_interactor.dart [hunk 3]
```diff
@@ -44,6 +44,9 @@ class GetAutoCompleteSharingInteractor {
       AutoCompleteType autoCompleteType) async {
     try {
       final resultList = await _autoCompleteRepository.getAutoComplete(autoCompletePattern, autoCompleteType);
+      if (resultList.isEmpty && autoCompletePattern.value.isValidEmail()) {
```
[the remark this answers is not in the document]
Comment (manhduc-nguyen) [reply domain/lib/src/usecases/autocomplete/get_autocomplete_sharing_interactor.dart:47]: if the input pattern is a valid email address and we don't have any results from server then this case is applied (for supporting other email domain like gmail.com, not only linshare.org)

**Comment by manhduc-nguyen** [comment]
@hoangdat @imGok @ngohado done, please check my fix up commit

File: lib/presentation/widget/upload_file/upload_file_viewmodel.dart [hunk 4]
```diff
@@ -69,8 +71,8 @@ class UploadFileViewModel extends BaseViewModel {
 
   Document _documentArgument;
 
-  final BehaviorSubject<List<String>> _userMailList = BehaviorSubject.seeded([]);
-  BehaviorSubject<List<String>> get userMailList => _userMailList;
+  final BehaviorSubject<List<AutoCompleteResult>> _userMailList = BehaviorSubject.seeded([]);
+  BehaviorSubject<List<AutoCompleteResult>> get userMailList => _userMailList;
```
[the remark this answers is not in the document]
Comment (ngohado) [reply lib/presentation/widget/upload_file/upload_file_viewmodel.dart]: I means the name could be `userMailsObservable`

File: lib/presentation/widget/upload_file/upload_file_widget.dart [hunk 5]
```diff
@@ -162,23 +171,59 @@ class _UploadFileWidgetState extends State<UploadFileWidget> {
...
+                      contentPadding: EdgeInsets.symmetric(vertical: 15),
+                      hintText: AppLocalizations.of(context).add_people,
+                      hintStyle: TextStyle(
+                          color: AppColor.uploadFileFileSizeTextColor,
+                          fontSize: 16.0),
+                      prefixIcon: Icon(
+                        Icons.person_add,
+                        size: 24.0,
+                      ))),
+              debounceDuration: Duration(milliseconds: 500),
```
Comment (hoangdat) [inline lib/presentation/widget/upload_file/upload_file_widget.dart:188]: should be 300 ms

File: lib/presentation/widget/upload_file/upload_file_widget.dart [hunk 6]
```diff
@@ -48,6 +51,7 @@ class _UploadFileWidgetState extends State<UploadFileWidget> {
   final uploadFileViewModel = getIt<UploadFileViewModel>();
   final imagePath = getIt<AppImagePaths>();
   final appNavigation = getIt<AppNavigation>();
+  final TextEditingController _typeAheadController = TextEditingController();
```
Comment (ngohado) [inline lib/presentation/widget/upload_file/upload_file_widget.dart:54]: I saw the cookbook of Flutter. I found that Flutter prefer a `TextEditingController` should be disposed after the holding widget is disposed. Plz consider https://flutter.dev/docs/cookbook/forms/text-field-changes

Comment (hoangdat) [reply lib/presentation/widget/upload_file/upload_file_widget.dart:54]: can you describe more about that

Comment (ngohado) [reply lib/presentation/widget/upload_file/upload_file_widget.dart:54]: Ive read the code of dispose, it just free out added listeners. In this our case, we don't have any listener, but with the responsible of a person who did at first place, I think we need to take care when any listeners will be added in the future :D

Comment (manhduc-nguyen) [reply lib/presentation/widget/upload_file/upload_file_widget.dart:54]: Agree, lets call dispose whenever use this.

**Review by hoangdat — APPROVED** [review]

**Review by ngohado — APPROVED** [review]

**Pull request merged**